File: app/src/main/java/com/renju_note/isoo/util/BoardLayout.java
package com.renju_note.isoo.util;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Point;
import android.graphics.Typeface;
import android.graphics.drawable.GradientDrawable;
import android.text.InputFilter;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.Display;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;

import com.renju_note.isoo.R;
import com.renju_note.isoo.data.SeqTree;

import org.jetbrains.annotations.NotNull;

public class BoardLayout extends ConstraintLayout {
    // for debuging (means coordinate that users add stones)
    private float debug_x, debug_y;

    // main paint for drawing lines
    private Paint paint;
    // board size by pixel
    public final int PX;
    // distance between dots
    public final float LINE;
    public final float DENSITY;

    // parent of board (MainActivity)
    private LinearLayout parent;

    // id for number coordinate
    int[] ids_coord_num;
    // id for word coordinate
    int[] ids_coord_char;
    // id for stone text view
    int[] ids_stone;
    // id for child text view
    int[] ids_blank;

    // boolean that coordinate exist
    boolean coord;

    // stones number + 1
    int sequence;

    SeqTree seqTree;

    // mode which is for touch event
    int mode = 0;

    // string that child's text
    String string;

    // part of basic settings
    String boardColor;
    String lineColor = "#666666";
    boolean sequenceVisible;
    int num_minus;

    // last stone's stroke color
    String lastStoneStrokeColor = "#D32560";

    public void setParent(LinearLayout L) {
        parent = L;
    }

    public BoardLayout(@NonNull @NotNull Context context) {
        this(context, null);
    }

    public BoardLayout(@NonNull @NotNull Context context, @Nullable @org.jetbrains.annotations.Nullable AttributeSet attrs) {
        super(context, attrs);

        WindowManager windowManager = (WindowManager)getContext().getSystemService(Context.WINDOW_SERVICE);
        Display display = windowManager.getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        if(size.x > 1200) {
            size.x = 1200;
        }
        int dp_margin = 10;
        DENSITY = density();
        int margin = (int) (dp_margin * DENSITY + 0.5);
        PX = size.x - margin;

        LINE = (float)(PX / 16);

        init();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        PX = w
    }

    private void init() {
        paint = new Paint();
        paint.setStrokeWidth((int)(PX / 351.3 + 0.5));

        ids_coord_num = new int[15];
        ids_coord_char = new int[15];
        ids_stone = new int[225];
        ids_blank = new int[225];

        sequence = 1;

        coord = false;

        seqTree = new SeqTree();

        SharedPreferences sf = getContext().getSharedPreferences("MainData", Activity.MODE_PRIVATE);
        boardColor = sf.getString("boardColor","#F2CA94");
        lineColor = sf.getString("lineColor","#666666");
        sequenceVisible = sf.getBoolean("sequenceVisible",true);
        num_minus = sf.getInt("num_minus",0);

        paint.setColor(Color.parseColor(lineColor));
    }

    // draw board line when application started
    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        for(int i = 1; i <= 15; i++) {
            canvas.drawLine(LINE*i+LINE/2,LINE-LINE/2,LINE*i+LINE/2,LINE*15-LINE/2,paint);
            canvas.drawLine(LINE+LINE/2,LINE*i-LINE/2, LINE*15+LINE/2,LINE*i-LINE/2,paint);
        }
        canvas.drawCircle(LINE*8+LINE/2,LINE*8-LINE/2,(int)(PX / 150.5 + 0.5),paint);
        canvas.drawCircle(LINE*4+LINE/2,LINE*12-LINE/2,(int)(PX / 150.5+ 0.5),paint);
        canvas.drawCircle(LINE*12+LINE/2,LINE*12-LINE/2,(int)(PX / 150.5+ 0.5),paint);
        canvas.drawCircle(LINE*4+LINE/2,LINE*4-LINE/2,(int)(PX / 150.5+ 0.5),paint);
        canvas.drawCircle(LINE*12+LINE/2,LINE*4-LINE/2,(int)(PX / 150.5+ 0.5),paint);

        canvas.drawLine(LINE/2,LINE*15,LINE/2,LINE*16,paint);
        canvas.drawLine(0,LINE*15+LINE/2,LINE,LINE*15+LINE/2,paint);
    }

    @SuppressLint({"SetTextI18n", "ClickableViewAccessibility"})
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float x, y;
        try {
            if (event.getAction() == MotionEvent.ACTION_UP) {
                if ((event.getX() + LINE / 2) % LINE <= LINE / 2)
                    x = event.getX() - ((event.getX() + LINE / 2) % LINE);
                else x = event.getX() - ((event.getX() + LINE / 2) % LINE) + LINE;

                if ((event.getY() - LINE / 2) % LINE <= LINE / 2)
                    y = event.getY() - ((event.getY() - LINE / 2) % LINE);
                else y = event.getY() - ((event.getY() - LINE / 2) % LINE) + LINE;
                // --------------------------------------------------------------------------- set x, y
                if ((int) ((x - LINE / 2) / LINE + 0.5) - 1 >= 0 && (int) ((y + LINE / 2) / LINE + 0.5) - 1 < 15) { // if x,y is in board
                    debug_x = ((x - LINE / 2) / LINE);
                    debug_y = ((y + LINE / 2) / LINE);
                    if (seqTree.getNow_board()[(int) ((x - LINE / 2) / LINE + 0.5) - 1][(int) ((y + LINE / 2) / LINE + 0.5) - 1] == 0) { // if x,y in board and there is no stone
                        //-----------------------------------------------------------------------
                        if (mode == 0) {
                            if (sequence != 1) {
                                if (sequence % 2 == 0)
                                    this.findViewById(ids_stone[sequence - 2]).setBackground(stoneDrawable(Color.BLACK,lineColor));
                                else
                                    this.findViewById(ids_stone[sequence - 2]).setBackground(stoneDrawable(Color.WHITE,lineColor));
                            }
                            // ----------------------------------------------------------------------- change last move drawable
                            addNumber(x, y, sequence);
                            sequence++;

                            seqTree.next(seqTree.getNow(), (int) ((x - LINE / 2) / LINE + 0.5) - 1, (int) ((y + LINE / 2) / LINE + 0.5) - 1, sequence);

                            int j = 0;
                            for (SeqTree.Node temp = seqTree.getNow().getParent().getChild(); temp != null; temp = temp.getNext()) {
                                this.removeViewInLayout(findViewById(ids_blank[j])); // remove blank
                                j++;
                            }

                            if (seqTree.getNow().getChild() != null) {
                                j = 0;
                                for (SeqTree.Node temp = seqTree.getNow().getChild(); temp != null; temp = temp.getNext()) {
                                    float c_x = (temp.getX() + 1) * LINE + LINE / 2;
                                    float c_y = (temp.getY() + 1) * LINE - LINE / 2;
                                    addChild(c_x, c_y, j, temp.getText());
                                    j++;
                                }
                            }

                            EditText editText = parent.findViewById(R.id.editText);
                            editText.setText(seqTree.getNow().getBoxText());
                        } else if (mode == 1) {

                            AlertDialog.Builder alert = new AlertDialog.Builder(getContext(), R.style.MyAlertDialogTheme2);
                            alert.setTitle("Put words");
                            alert.setMessage("The maximum number of characters is 3.\n\nThere is a risk of missing text if it is too long.\n");
                            final EditText word = new EditText(getContext());
                            word.setGravity(Gravity.CENTER);
                            word.setHint("Put words here");
                            InputFilter[] FilterArray = new InputFilter[1];
                            FilterArray[0] = new InputFilter.LengthFilter(3);
                            word.setFilters(FilterArray);
                            alert.setView(word);
                            alert.setPositiveButton("Ok", (dialog, whichButton) -> {
                                string = word.getText().toString();
                                int c_x = (int) ((x - LINE / 2) / LINE + 0.5) - 1;
                                int c_y = (int) ((y + LINE / 2) / LINE + 0.5) - 1;

                                if (seqTree.getNow().getChild() == null) {
                                    seqTree.createChild(seqTree.getNow(), c_x, c_y);
                                    seqTree.getNow().getChild().setText(string);
                                    addChild(x, y, 0, seqTree.getNow().getChild().getText());
                                } else {
                                    int j = 0;
                                    for (SeqTree.Node temp = seqTree.getNow().getChild(); temp != null; temp = temp.getNext()) {
                                        if (temp.getX() == c_x && temp.getY() == c_y) {
                                            temp.setText(string);
                                            ((TextView) (findViewById(ids_blank[j]))).setText(temp.getText());
                                            break;
                                        }
                                        if (temp.getNext() == null) {
                                            seqTree.createNext(temp, seqTree.getNow(), c_x, c_y);
                                            temp.getNext().setText(string);
                                            addChild(x, y, j + 1, temp.getNext().getText());
                                            break;
                                        }
                                        j++;
                                    }
                                }
                            });
                            alert.setNegativeButton("Cancel", (dialog, whichButton) -> {
                            });
                            alert.show();

                        }
                        //------------------------------------------------------------------------
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();

            TextView debug = parent.findViewById(R.id.debug);
            debug.setText("developer mode\n"+" / "+ debug_x +" / "+ debug_y +" / "+ PX +" / "+ LINE);
        }
        TextView seq = parent.findViewById(R.id.sequence);
        seq.setText(String.valueOf(sequence-1));
        return super.onTouchEvent(event);
    }

    // set size of board by depending on the size of machine
    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        setMeasuredDimension(PX,PX);
    }

    public static int generateViewId() {
        return View.generateViewId();
    }

    private float density() {
        DisplayMetrics displayMetrics = new DisplayMetrics();
        ((Activity)getContext()).getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        return displayMetrics.density;
    }

    public GradientDrawable stoneDrawable(int stoneColor, String strokeColor) {
        GradientDrawable drawable = new GradientDrawable();
        if(stoneColor == Color.parseColor(boardColor) + 10000) {
            drawable.setColor(stoneColor - 10000);
        }
        else {
            drawable.setStroke((int) (PX / 351.3 + 0.5), Color.parseColor(strokeColor));
            drawable.setColor(stoneColor);
        }
        drawable.setShape(GradientDrawable.OVAL);
        return drawable;
    }

    // draw board line when user changed a line color
    public void drawLine(Canvas canvas) {
        super.draw(canvas);
        paint.setColor(Color.parseColor(lineColor));

        for(int i = 1; i <= 15; i++) {
            canvas.drawLine(LINE*i+LINE/2,LINE-LINE/2,LINE*i+LINE/2,LINE*15-LINE/2,paint);
            canvas.drawLine(LINE+LINE/2,LINE*i-LINE/2, LINE*15+LINE/2,LINE*i-LINE/2,paint);
        }
        canvas.drawCircle(LINE*8+LINE/2,LINE*8-LINE/2,(int)(PX / 150.5 + 0.5),paint);
        canvas.drawCircle(LINE*4+LINE/2,LINE*12-LINE/2,(int)(PX / 150.5 + 0.5),paint);
        canvas.drawCircle(LINE*12+LINE/2,LINE*12-LINE/2,(int)(PX / 150.5 + 0.5),paint);
        canvas.drawCircle(LINE*4+LINE/2,LINE*4-LINE/2,(int)(PX / 150.5 + 0.5),paint);
        canvas.drawCircle(LINE*12+LINE/2,LINE*4-LINE/2,(int)(PX / 150.5 + 0.5),paint);

        canvas.drawLine(LINE/2,LINE*15,LINE/2,LINE*16,paint);
        canvas.drawLine(0,LINE*15+LINE/2,LINE,LINE*15+LINE/2,paint);
    }
}

File: app/src/main/java/com/renju_note/isoo/util/BoardLayout.kt
package com.renju_note.isoo.util

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.constraintlayout.widget.ConstraintLayout
import com.renju_note.isoo.data.BoardColorSetting
import com.renju_note.isoo.data.BoardDisplaySetting
import java.io.Serializable
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin

class BoardLayout(context: Context, attrs: AttributeSet?) : ConstraintLayout(context, attrs), Serializable {

    private var lineInterval : Float = 0f
    private var density : Float = 0f
    private val paint = Paint()
    private val bounds = Rect()

    private var boardColorSetting = BoardColorSetting.getDefaultSetting()
    private var boardDisplaySetting = BoardDisplaySetting.getDefaultSetting()

    private val stones = HashMap<String, StoneView>()
    private val childs = ArrayList<Child>()
    private val drawingElements = ArrayList<Element>() // Line, Area, Arrow
    private val points = ArrayList<Point>()

    interface Element {
        fun draw(canvas : Canvas?)
    }

    inner class Line(private val startX : Int, private val startY : Int, private val endX : Int, private val endY : Int) : Element {
        private val lineColor = boardColorSetting.drawLineColor

        override fun draw(canvas : Canvas?) {
            paint.apply {
                color = Color.parseColor(lineColor)
                style = Paint.Style.STROKE
                strokeWidth = lineInterval / 9.4107f
                pathEffect = DashPathEffect(floatArrayOf(10f, 5f), 0f)
            }
            canvas?.drawLine(getRealX(startX), getRealY(startY), getRealX(endX), getRealY(endY), paint)
        }

        fun hasNoLength() : Boolean {
            return (startX == endX) && (startY == endY)
        }
    }

    inner class Area(private var left : Int, private var top : Int, private var right : Int, private var bottom : Int) : Element {
        private val areaColor = boardColorSetting.drawAreaColor

        override fun draw(canvas : Canvas?) {
            paint.apply {
                color = Color.parseColor(areaColor)
                style = Paint.Style.FILL
            }
            if(left > right) {
                val tmp = left
                left = right
                right = tmp
            }
            if(top > bottom) {
                val tmp = top
                top = bottom
                bottom = tmp
            }
            val margin = lineInterval / 2
            val rect = RectF(getRealX(left) - margin, getRealY(top) - margin, getRealX(right) + margin, getRealY(bottom) + margin)
            val path = Path()
            path.addRoundRect(rect, lineInterval / 1.6468f, lineInterval / 1.6468f, Path.Direction.CW)
            canvas?.drawPath(path, paint)
        }
    }

    inner class Arrow(private val startX : Int, private val startY : Int, private val endX : Int, private val endY : Int) : Element {
        private val arrowColor = boardColorSetting.drawArrowColor

        override fun draw(canvas: Canvas?) {
            val width = lineInterval / 2
            val startX = getRealX(this.startX)
            val startY = getRealY(this.startY)
            val endX = getRealX(this.endX)
            val endY = getRealY(this.endY)
            val angle = atan2((endX - startX), (endY - startY))
            val angle90 = angle - Math.toRadians(90.0)

            val p1 = Pair(startX + width/2*sin(angle90), startY + width/2*cos(angle90))
            val p2 = Pair(startX - width/2*sin(angle90), startY - width/2*cos(angle90))
            val p3 = Pair(endX - sin(angle)*width + sin(angle90)*width/2,
                endY - cos(angle)*width + cos(angle90)*width/2)
            val p4 = Pair(endX - sin(angle)*width + sin(angle90)*width,
                endY - cos(angle)*width + cos(angle90)*width)
            val p5 = Pair(endX - sin(angle)*width - sin(angle90)*width/2,
                endY - cos(angle)*width - cos(angle90)*width/2)
            val p6 = Pair(endX - sin(angle)*width - sin(angle90)*width,
                endY - cos(angle)*width - cos(angle90)*width)

            val path = Path()
            path.moveTo(p1.first.toFloat(), p1.second.toFloat())
            path.lineTo(p2.first.toFloat(), p2.second.toFloat())
            path.lineTo(p5.first.toFloat(), p5.second.toFloat())
            path.lineTo(p6.first.toFloat(), p6.second.toFloat())
            path.lineTo(endX, endY)
            path.lineTo(p4.first.toFloat(), p4.second.toFloat())
            path.lineTo(p3.first.toFloat(), p3.second.toFloat())
            path.lineTo(p1.first.toFloat(), p1.second.toFloat())
            path.close()

            paint.style = Paint.Style.FILL
            paint.pathEffect = null
            paint.color = Color.parseColor(arrowColor)
            canvas?.drawPath(path, paint)
        }

        fun hasNoLength() : Boolean {
            return (startX == endX) && (startY == endY)
        }
    }

    inner class Point(private val x : Int, private val y : Int, private val pointColor : String) : Element {
        override fun draw(canvas : Canvas?) {
            paint.apply {
                color = Color.parseColor(pointColor)
                style = Paint.Style.FILL
                pathEffect = null
            }
            canvas?.drawCircle(getRealX(x), getRealY(y), lineInterval / 6.5875f, paint)
        }
    }

    enum class StoneViewType { BLACK, WHITE }

    inner class StoneView(private var type : StoneViewType, private var index : Int, private val x : Int, private val y : Int) : Element {
        val id = "$x/$y"

        fun setStoneType(type : StoneViewType) {
            this.type = type
        }

        fun getStoneType() : StoneViewType = type

        fun setStoneIndex(idx : Int) {
            index = idx
        }

        fun getStoneIndex() : Int = index

        override fun draw(canvas: Canvas?) {
            val radius = lineInterval / 2 - 2
            paint.apply {
                pathEffect = null
                strokeWidth = (width / 351.3 + 0.5).toInt().toFloat()
                color = if(index == stones.size)
                    Color.parseColor(blendColors("#FFFFFF", boardColorSetting.lastStoneStrokeColor))
                else
                    Color.parseColor(blendColors("#FFFFFF", boardColorSetting.lineColor))
                style = Paint.Style.STROKE
            }
            canvas?.drawCircle(getRealX(x), getRealY(y), radius, paint)

            paint.style = Paint.Style.FILL
            if(type == StoneViewType.BLACK)
                paint.color = Color.BLACK
            else
                paint.color = Color.WHITE
            canvas?.drawCircle(getRealX(x), getRealY(y), radius - paint.strokeWidth / 2, paint)

            if(boardDisplaySetting.sequenceVisible && index - boardDisplaySetting.startPoint > 0) {
                paint.apply {
                    color = if (type == StoneViewType.BLACK)
                        Color.WHITE
                    else
                        Color.BLACK
                    textSize = lineInterval / 2.2f
                    typeface = Typeface.create(Typeface.MONOSPACE, Typeface.BOLD)
                }
                val idxText = (index - boardDisplaySetting.startPoint).toString()
                paint.getTextBounds("333", 0, idxText.length, bounds)
                val centerX = bounds.exactCenterX()
                val centerY = bounds.exactCenterY()
                canvas?.drawText(idxText, getRealX(x) - centerX, getRealY(y) - centerY, paint)
            }
        }
    }

    inner class Child(private var text : String, private val x : Int, private val y : Int) : Element {
        override fun draw(canvas: Canvas?) {
            if(!boardDisplaySetting.nextNodeVisible) return
            paint.pathEffect = null
            paint.strokeWidth = (width / 351.3 + 0.5).toInt().toFloat()
            paint.style = Paint.Style.FILL
            if (text.isNotEmpty()) {
                paint.color = Color.parseColor(blendColors("#FFFFFF", boardColorSetting.boardColor))
                paint.getTextBounds(text, 0, text.length, bounds)
                if(bounds.width() > lineInterval) {
                    text = text.substring(0, text.length - 1)
                    paint.getTextBounds(text, 0, text.length, bounds)
                }
                val centerX = bounds.exactCenterX()
                val centerY = bounds.exactCenterY()
                canvas?.drawRect(
                    getRealX(x) - centerX,
                    getRealY(y) - centerY + lineInterval / 13.2f,
                    getRealX(x) + centerX,
                    getRealY(y) + centerY - lineInterval / 13.2f, paint
                )
                paint.color = Color.parseColor(blendColors("#FFFFFF", boardColorSetting.textColor))
                paint.textSize = lineInterval / 2.2f
                paint.typeface = Typeface.create(Typeface.MONOSPACE, Typeface.BOLD)
                canvas?.drawText(
                    text,
                    getRealX(x) - centerX,
                    getRealY(y) - centerY,
                    paint
                )
            } else {
                paint.color = Color.parseColor(blendColors("#FFFFFF", boardColorSetting.nodeColor))
                canvas?.drawCircle(getRealX(x), getRealY(y), lineInterval / 34 * 6, paint)
            }
        }
    }

    open class OnBoardTouchListener {
        open fun getCoordinates(x : Int, y : Int) {
            return
        }
    }

    private var onBoardTouchListener = OnBoardTouchListener()

    fun setOnBoardTouchListener(listener : OnBoardTouchListener) {
        onBoardTouchListener = listener
    }

    init {
        setBackgroundColor(Color.TRANSPARENT)
        density = context.resources.displayMetrics.density
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        updateBoardStatus(boardColorSetting, boardDisplaySetting)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        lineInterval = MeasureSpec.getSize(widthMeasureSpec).toFloat() / 16
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        paint.isAntiAlias = true
        drawBoard(canvas)
        drawingElements.forEach { drawingElement ->
            if(drawingElement is Line) drawingElement.draw(canvas)
        }
        childs.forEach { child -> child.draw(canvas) }
        stones.keys.forEach { key -> stones[key]?.draw(canvas) }
        drawingElements.forEach { drawingElement ->
            if(drawingElement is Arrow) drawingElement.draw(canvas)
        }
        drawingElements.forEach { drawingElement ->
            if(drawingElement is Area) drawingElement.draw(canvas)
        }
        points.forEach { point -> point.draw(canvas) }
    }

    private fun drawBoard(canvas : Canvas?) {
        // board background
        paint.pathEffect = null
        paint.style = Paint.Style.FILL
        paint.color = Color.parseColor(boardColorSetting.boardColor)
        canvas!!.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)

        // board lines
        paint.color = Color.parseColor(blendColors("#FFFFFF", boardColorSetting.lineColor))
        paint.strokeWidth = (width / 351.3 + 0.5).toInt().toFloat()
        val x = lineInterval
        for (i in 1..15) {
            canvas.drawLine(x * i + x / 2, x - x / 2, x * i + x / 2, x * 15 - x / 2, paint)
            canvas.drawLine(x + x / 2, x * i - x / 2, x * 15 + x / 2, x * i - x / 2, paint)
        }
        canvas.drawCircle(x * 8 + x / 2, x * 8 - x / 2, (width / 150.5 + 0.5).toInt().toFloat(), paint)
        canvas.drawCircle(x * 4 + x / 2, x * 12 - x / 2, (width / 150.5 + 0.5).toInt().toFloat(), paint)
        canvas.drawCircle(x * 12 + x / 2, x * 12 - x / 2, (width / 150.5 + 0.5).toInt().toFloat(), paint)
        canvas.drawCircle(x * 4 + x / 2, x * 4 - x / 2, (width / 150.5 + 0.5).toInt().toFloat(), paint)
        canvas.drawCircle(x * 12 + x / 2, x * 4 - x / 2, (width / 150.5 + 0.5).toInt().toFloat(), paint)

        canvas.drawLine(x / 2, x * 15, x / 2, x * 16, paint)
        canvas.drawLine(0f, x * 15 + x / 2, x, x * 15 + x / 2, paint)

        // board coordinate
        paint.textSize = lineInterval / 2.2f
        paint.typeface = Typeface.MONOSPACE
        for(i in 0..14) {
            val mText = (15 - i).toString()
            paint.getTextBounds("45", 0, mText.length, bounds)
            val textWidth = bounds.width()
            val textHeight = bounds.height()
            canvas.drawText(mText, x / 2 - textWidth / 2, x * i + x / 2 + textHeight / 2, paint)
        }
        for(i in 0..14) {
            val mText = (i + 65).toChar().toString()
            paint.getTextBounds("A", 0, mText.length, bounds)
            val textWidth = bounds.width()
            val textHeight = bounds.height()
            canvas.drawText(mText, x * (i + 1) + x / 2 - textWidth / 2, x * 15 + x / 2 + textHeight / 2, paint)
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val x: Float
        val y: Float
        try {
            if (event.action == MotionEvent.ACTION_UP) {
                x = if ((event.x + lineInterval / 2) % lineInterval <= lineInterval / 2) event.x - (event.x + lineInterval / 2) % lineInterval else event.x - (event.x + lineInterval / 2) % lineInterval + lineInterval
                y = if ((event.y - lineInterval / 2) % lineInterval <= lineInterval / 2) event.y - (event.y - lineInterval / 2) % lineInterval else event.y - (event.y - lineInterval / 2) % lineInterval + lineInterval
                if (((x - lineInterval / 2) / lineInterval + 0.5).toInt() - 1 >= 0 && ((y + lineInterval / 2) / lineInterval + 0.5).toInt() - 1 < 15) { // if x,y is in board
                    val xc = ((x - lineInterval / 2) / lineInterval + 0.5).toInt() - 1
                    val yc = ((y + lineInterval / 2) / lineInterval + 0.5).toInt() - 1
                    onBoardTouchListener.getCoordinates(xc, yc)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return true
    }

    fun placeStones(addStones : ArrayList<StoneView>?, deleteStonesID : ArrayList<String>?, childs : ArrayList<Child>) {
        if (deleteStonesID != null) {
            for (deleteStoneID in deleteStonesID) {
                stones.remove(deleteStoneID)
            }
        }

        if (addStones != null) {
            for(addStone in addStones) {
                if(stones.contains(addStone.id)) {
                    stones[addStone.id]!!.setStoneIndex(addStone.getStoneIndex())
                    stones[addStone.id]!!.setStoneType(addStone.getStoneType())
                } else {
                    stones[addStone.id] = addStone
                }
            }
        }

        this.childs.clear()
        this.childs.addAll(childs)
        invalidate()
    }

    fun removeAllStones() {
        stones.clear()
        childs.clear()
        invalidate()
    }

    fun addLine(line : Line) {
        if(line.hasNoLength()) return
        drawingElements.add(line)
        invalidate()
    }

    fun addArea(area : Area) {
        drawingElements.add(area)
        invalidate()
    }

    fun addArrow(arrow : Arrow) {
        if(arrow.hasNoLength()) return
        drawingElements.add(arrow)
        invalidate()
    }

    fun addPoint(point : Point) {
        points.add(point)
        invalidate()
    }

    fun deleteLastDrawingElement() {
        points.clear()
        if(drawingElements.isNotEmpty())
            drawingElements.removeLast()
        invalidate()
    }

    fun deleteAllDrawingElements() {
        drawingElements.clear()
        points.clear()
        invalidate()
    }

    fun deleteAllPoints() {
        points.clear()
        invalidate()
    }

    fun getStoneID(x : Int, y : Int) : String = "$x/$y"

    private fun getRealX(x : Int) : Float = (x + 1) * lineInterval + lineInterval / 2

    private fun getRealY(y : Int) : Float = (y + 1) * lineInterval - lineInterval / 2

    fun updateBoardStatus(boardColorSetting : BoardColorSetting, boardDisplaySetting : BoardDisplaySetting) {
        this.boardColorSetting = boardColorSetting
        this.boardDisplaySetting = boardDisplaySetting
        invalidate()
    }

    private fun blendColors(baseColor: String, blendColor: String): String {
        val base = Color.parseColor(baseColor)
        val blend = Color.parseColor(blendColor)
        val alpha = Color.alpha(blend) / 255f
        val red = (1 - alpha) * Color.red(base) + alpha * Color.red(blend)
        val green = (1 - alpha) * Color.green(base) + alpha * Color.green(blend)
        val blue = (1 - alpha) * Color.blue(base) + alpha * Color.blue(blend)
        val alphaInt = (alpha * 255).toInt()
        val colorInt = Color.argb(alphaInt, red.toInt(), green.toInt(), blue.toInt())
        return String.format("#%06X", 0xFFFFFF and colorInt)
    }

}